src/Pirouette/SMT/Constraints.hs

```diff
@@ -25,50 +25,45 @@ import Pirouette.Term.Syntax
 import qualified Pirouette.Term.Syntax.SystemF as SystF
 import Prettyprinter hiding (Pretty (..))
 import qualified PureSMT
+import qualified UnionFind as UF
 
 data ConstraintSet lang meta = ConstraintSet
-  { -- | Represents the set of productive symbolic variable assignments we know of.
-    -- Productive in the sense that we will always have at least one constructor
-    -- on the rhs.
-    csAssignments :: M.Map meta (TermMeta lang meta),
-    -- | Represents the symbolic variable equivalences we know of. Satisfies the invariant
-    -- that if @M.lookup v csmetaEq == Just v'@, then @v' < v@ (lexicographically)
-    csMetaEq :: M.Map meta meta,
+  { -- | Represents the set of productive symbolic variable assignments we know
+    -- of. Productive in the sense that we will always have at least one
+    -- constructor on the rhs. The assignments are stored from /equivalence
+    -- classes of symbolic variables/ to @TermMeta@.
+    csAssignments :: UF.UnionFind meta (TermMeta lang meta),
     -- | Any potential native constraint we might need.
     csNative :: [PureSMT.SExpr],
     -- | And finally, pairs of terms we've discovered /not/ to be equal
     csRelations :: [(TermRelation, TermMeta lang meta, TermMeta lang meta)]
   }
 
-instance (Ord meta) => Semigroup (ConstraintSet lang meta) where
-  s1 <> s2 =
-    ConstraintSet
-      (csAssignments s1 <> csAssignments s2)
-      (csMetaEq s1 <> csMetaEq s2)
-      (csNative s1 <> csNative s2)
-      (csRelations s1 <> csRelations s2)
+-- instance (Ord meta, Semigroup meta) => Semigroup (ConstraintSet lang meta) where
+--   s1 <> s2 =
+--     ConstraintSet
+--       (csAssignments s1 <> csAssignments s2)
```

Comment:
This would need to be implemented for Union-Find, probably by requiring the values (here, `TermMeta lang meta`) to be a semigroup as well. Otherwise, we end up with something potentially buggy, as I think is the case in #146 (cf https://github.com/tweag/pirouette/pull/146#discussion_r964637039).